packages/token-sdk/src/repository.ts

```diff
@@ -0,0 +1,140 @@
+import { Address } from "@orca-so/common-sdk";
+import { Mintlist, Token } from "./types";
+import { TokenFetcher } from "./fetcher";
+
+/**
+ * Token with tags. Tags are strings that can be used to label tokens that do not exist on-chain.
+ */
+export type TokenWithTags = Token & { tags: string[] };
+
+/**
+ * Repository for managing a local instance of token mints and fetching token metadata.
+ *
+ * Mints can also be tagged with a string. This is useful when groups of mints need to be handled
+ * together. For example, a UI may want to mark all mints labeled with a "whitelisted" tag.
+ *
+ * All mints managed by an instance of this class can be retrieved to represent all tokens in the
+ * local state.
+ */
+export class TokenRepository {
+  private readonly mintMap: Map<string, string[]> = new Map();
+  private readonly tagMap: Map<string, string[]> = new Map();
+  private readonly excluded: Set<string> = new Set();
+
+  constructor(private readonly fetcher: TokenFetcher) {}
+
+  /**
+   * Adds a mint to the repository. If the mint already exists, the tags are merged.
+   * @param mint Mint to add
+   * @param tags Tags to add to the mint
+   * @returns This instance of the repository
+   */
+  addMint(mint: Address, tags: string[] = []): TokenRepository {
+    const mintString = mint.toString();
+    const tagSet = new Set(this.mintMap.get(mintString));
+    tags.forEach((tag) => tagSet.add(tag));
+    this.mintMap.set(mintString, Array.from(tagSet));
+
+    tags.forEach((tag) => {
+      const mintSet = new Set(this.tagMap.get(tag));
+      mintSet.add(mintString);
+      this.tagMap.set(tag, Array.from(mintSet));
+    });
+    return this;
+  }
+
+  /**
+   * Adds multiple mints to the repository. If a mint already exists, the tags are merged. Tags are
+   * added to all mints.
+   * @param mints Mints to add
+   * @param tags Tags to add to the mints
+   * @returns This instance of the repository
+   */
+  addMints(mints: Address[], tags: string[] = []): TokenRepository {
+    mints.forEach((mint) => this.addMint(mint, tags));
```

Comment:
hmm, is it necessary to recreate the ``tagMap`` len(mints) times?
I think recreate tagMap part can be isolated method like "buildTagMap(tag)"

Reply:
I changed the implementation to use `addMints` as the base method for adding mints so its done in bulk. Also changed the map values to use sets instead of arrays of strings. I believe this makes it more efficient and addresses your comment. Let me know if that looks good to you :)

Reply:
Very nice, ``Map<string, Set>`` looks efficient. 🙂